src/main/scala/lantern/TensorDifferentiation.scala

```diff
@@ -993,6 +997,14 @@ trait TensorDsl extends DslOps with Diff {
         offset += input.x.shape.strides(0)
       }
     }
+
+    override def sum(x: Tensor): Tensor = {
+      Tensor.scalar(x.fold(0.0f)(_ + _))
+    }
+
+    override def sum_grad(input: TensorR, res: TensorR): Unit = {
+      input.d += res.d
```

Comment:
This should really be `+=(input.d, res.d)`. [Calling methods defined on `Tensor` is dangerous because it may dispatch to a different backend.](https://github.com/feiwang3311/Lantern/pull/35/files#r227529551)